File: 131-palindrome-partitioning/main.py
from typing import List


class Solution:
	def partition(self, s: str) -> List[List[str]]:
		res = []

		# current partition
		part = []

		def dfs(i):
			# Since this is a recursive function, first thing we do is to check the base case
			if i >= len(s):
				# When we get to here, we know that we have a valid partition and we have no more characters to add, so let's add
				# it to the res
				res.append(part.copy())
				print("hello: ", part)

				return

			# with this loop, we're generating every single possible substring from i to j and if it's a palindrome, we add it to `part`.
			# If it's not a palindrome, we skip it using the if block
			for j in range(i, len(s)):
				# is `s` starting from i and ending in j, a palindrome?
				if self.isPali(s, i, j):
					# Note: We wanna include j as well, because sub-listing in python is exclusive, we need to add 1 to j
					part.append(s[i:j+1])
					print("ok: ", i, j , s[i:j+1])

					dfs(j + 1)

					# after doing the recursive calls, we have to clean up:
					part.pop()

		dfs(0)

		return res

	def isPali(self, s, l, r):
		while l < r:
			if s[l] != s[r]:
				return False

			l, r = l + 1, r - 1

		return True

File: 131-palindrome-partitioning/main.js
const isPalindrome = (s, start, end) => {
    while (start < end) {
        if (s[start] !== s[end]) return false

        start++
        end--
    }

    return true
}

/**
 * @param {string} s
 * @return {string[][]}
 */
var partition = function(s) {
    // global result
    const result = []

    // dfs recursive helper
    const dfs = (i, slate) => {
        // backtracking case(we don't need in this case)

        // base case. We're at the leaf level of the tree
        if (i === s.length) {
            result.push(slate.slice())

            return
        }

        // dfs recursive case
        /* Note: Yeah we're doing DFS, but we use for loop to do breadth first on current level. We're going DFS to the leaf,
        but at the root level, we have multiple choices to go, hence the for loop. */
        for (let j = i; j < s.length; j++) {
            // here, we have to do some checking to whether we wanna go further down the recursive tree or not
            // if the range between i and j is a palindrome, then we wanna add the range to the slate
            if (isPalindrome(s, i, j)) {
                // slice is not inclusive from the end, that's why we use `j + 1`
                slate.push(s.slice(i, j + 1))

                dfs(j + 1, slate)

                slate.pop()
            }
        }
    }

    dfs(0, [])

    return result
};